Document/0x07a-Testing-Authentication-and-Session-Management.md

```diff
@@ -66,6 +66,98 @@ If any of these two conditions raise an issue, reject the request and do not all
+```JSON
+HMACSHA256(
+  base64UrlEncode(header) + "." +
+  base64UrlEncode(payload),
+  secret
+)
+```
+
+JWT implementations are available for all major programming languages, like PHP<sup>[2]</sup> or Java Spring<sup>[3]</sup>.
+
+#### Static Analysis
+
+Identify the JWT library that is used on server and client side. It should be checked if there are any known vulnerabilities available for the JWT libraries in use.
+
+The following best practices should be verified in the JWT libraries<sup>[7]</sup>:
+* Verify the signature on server-side at all times for all incoming requests containing a token.
+* Verify the data embedded into the JWT and check if sensitive information is available. If so, it should be considered to encrypt the JWT content.
+* Verify if replay attacks are addressed by using `jti` (JWT ID) claim, which provides a unique identifier for JWT.
+* Verify where the secret signing key is located and stored. It should only be available for the issuer and consumer.
+
+#### Dynamic Analysis
+
+Several known vulnerabilities in JWT should be checked while executing a dynamic analysis:
+* Hashing algorithm `none`<sup>[6]</sup>:
+  * Modify the `alg` attribute in the token header and delete `HS256` and set it to `none` and use an empty signature (e.g. signature = ""). Use this token and replay it in a request. Some libraries treat tokens signed with the none algorithm as a valid token with a verified signature. This would allow an attacker to create their own "signed" tokens.
+* Usage of asymmetric algorithms:
+  * Within JWT asymmetric algorithms as RSA or ECDSA can be used for signing the token. Due to vulnerabilities in some JWT frameworks the public key can be misused as HMAC secret key in order to sign the token.
+* Token Storage on client side:
+  * When using a mobile app that uses JWT it should be verified where the token is stored locally on the device<sup>[5]</sup>.
+* Information Disclosure:
+  * Decode the Base-64 encoded JWT and check what kind of data is transmitted within it.
+
+Please also follow the test cases in the OWASP JWT Cheat Sheet<sup>[4]</sup> if JWT is used.
+
+#### Remediation
+
+The latest version available of the JWT libraries in use should be implemented, to avoid known vulnerabilities.
+
+Store the JWT on the mobile phone using a secure mechanism, like KeyChain on iOS or KeyStore on Android.
+
+If replay attacks are a risk for the app, `jti` (JWT ID) claim should be implemented.
+
+Ideally the content of JWT should be encrypted. An example implementation in Java can be found in the OWASP JWT Cheat Sheet<sup>[4]</sup>
+
+#### References
+
+##### OWASP Mobile Top 10 2016
+
+* M4 - Insecure Authentication - https://www.owasp.org/index.php/Mobile_Top_10_2016-M4-Insecure_Authentication
+
+##### OWASP MASVS
+
+* 4.1: "If the app provides users with access to a remote service, an acceptable form of authentication such as username/password authentication is performed at the remote endpoint."
```

Comment:
I don't know. It is sort of an alternative for 4.2. However, if you use a JWT instead of a session token you are violating 4.2, "the remote endpoint uses randomly generated access tokens".